### ISSUE TITLE
STUN adress?

### ISSUE THREAD

**Issue by Thanh-Binh** [ultravideo/kvazzup#4]
I got an error by running like:
NetworkCandidates:                      Warning! Either couldn't find interface or this interface has run out of ports
NetworkCandidates:                      BUG: Not port set. Can't get STUN address

But really do not know how to avoid it.
Could you pls help me? Thanks

**Comment by polusto** [comment]
Are you connected to the internet? If not you can probably ignore the error.

Does it cause the call to not start?

**Comment by Thanh-Binh** [comment]
yes. it use WLAN for communicating with other PC.
I see, it rings by another computer

**Comment by Thanh-Binh** [comment]
In one laptop, I got an error like that it "failed to add camera. Does it have supported formats"?
Any special camera requirement?

**Comment by polusto** [comment]
Can you paste the whole log form both computers?

**Comment by Thanh-Binh** [comment]
After enabling camera in another PC working, I do not see STUN problems, but it crashed in both apps with the following error:
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
Filter:                                 Parameter set found (Frame received before: 0)
Filter:                                 PEER ERROR:Detected no slices in incoming stream.
Filter:                                 Uniniating.
[hevc @ 0x55638ed99ff0] flush, decoder_0.
Filter:                                 Starting to initiate OpenHEVC
[hevc @ 0x55638ede0b40] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
[hevc @ 0x55638ede0fd0] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
Filter:                                 OpenHEVC initiation successful. (Version: OpenHEVC v2.0)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioCaptureFilter:                     Audio Input State changed (States:: Current: 0, Wanted: 0)
TCPConnection:                          Starting to connect TCP
TCPConnection:                          Checking status (Address: 192.168.178.111:5060)
TCPConnection:                          Attempting to connect (Attempt: 1)
TCPConnection:                          Warning! TCP socket disconnected
QObject: Cannot create children for a parent that is in a different thread.
(Parent is QNativeSocketEngine(0x7f3a98004a90), parent's thread is TCPConnection(0x55638e7ac620), current thread is QThread(0x55638db62860)
TCPConnection:                          Connected succesfully (Connection: 192.168.178.101:41600 <-> 192.168.178.111:5060)
QSocketNotifier: Socket notifiers cannot be enabled or disabled from another thread
QSocketNotifier: Socket notifiers cannot be enabled or disabled from another thread
Segmentation fault (core dumped)

**Comment by polusto** [comment]
I pushed few fix commits to master from another branch related to TCP Connection. At least it hopefully shouldn't crash. Can you try again with newest master and paste the log?

Have you tried calling yourself on localhost (127.0.0.1)?

**Comment by Thanh-Binh** [comment]
Thanks.
Does it work when I call myself on local host? Because both apps access on the same built-in camera...

**Comment by polusto** [comment]
Just have one application running and call the same application. You can then answer the call yourself and you will see two camera feeds.

**Comment by Thanh-Binh** [comment]
in my test with 2 instances, by running the 2.instance I got error like
 ============================================================================= 
VideoDrawHelper:                        Warning! Tried updating target rect before picture
CameraBin error: "Device '/dev/video0' is busy"
GUIMessage:                             Closing message (Waiting messages: 0)

**Comment by polusto** [comment]
I meant, just use one instance. Two instances doesn't work, but one instance does.

**Comment by Thanh-Binh** [comment]
understand.

**Comment by Thanh-Binh** [comment]
I tested self-calling with current SW, and have STUN problem:
NetworkCandidates:                      BUG: Not port set. Can't get STUN address.

**Comment by polusto** [comment]
Can you paste the whole log?

**Comment by Thanh-Binh** [comment]
I have just updated your modified version, and tested.
It works well by self-calling.
But two camera feeds are flipped (I am over head).
Pls note my small camera video (bottom right) is correct.
How can I correct it? THank

**Comment by polusto** [comment]
I added a way to fix this to settings in master. Enable Settings -> Video settings -> Other -> Force call video vertical flip

This should flip the video streams so they are oriented in correct way.

**Comment by Thanh-Binh** [comment]
I will check it soon, thanks

**Comment by Thanh-Binh** [comment]
But I really do not understand why does image flipped?
Can you explain me?

**Comment by Thanh-Binh** [comment]
I think the flip-problem should be due to condertion RGB32toYUV and back!
What do you think?

**Comment by polusto** [comment]
Unfortunately at the moment I don't have time to investigate this further at this time. 

You are right that one of the conversion optimizations does flip the image and that is probably a bug. What I don't understand is why this flipping only occurs on Linux and not on Windows so that is why I'm afraid to just flip the image blindly somewhere.

Currently I'm the only one developing Kvazzup and my current goal is to get version of Kvazzup finished that is easily usable by anyone. Once that is done, I'll start looking more into specific issues like this.

**Comment by Thanh-Binh** [comment]
Understand!
There are also some bug by suddenly stop conversation, the GUI hangs and can not be closed! 
let me have time for cleaning my codes about CustomCamera template and it’s integration into your framework and update it soon

**Comment by polusto** [comment]
If GUI hangs, then the main thread is blocked. Qt works in such a way that you always have to use a separate thread when processing stuff continuously or in other words, not block the main thread.

**Comment by Thanh-Binh** [comment]
Thanks for your hints on thread.
About Video Compression:
I understood that we have to 
1. convert camera RGB-images into YUV format using RGB32toYUV 
2. then do compress algo on YUV-image
3. send ouput of video compression 
I am not familiar to video compression, do that I do not know how does the output of video compression look like? An array of bytes with the given length?
Can you pls explain me?

**Comment by polusto** [comment]
Kvazzup uses Kvazaar HEVC encoder for compression. Kvazaar takes in YUV video. If you output RGB32 video, it has to be converted to YUV. Kvazzup should already insert conversion filter when needed if the camera filter is constructed correctly (output is set to RGB32VIDEO).

Compressed video is a bitstream that is specified by HEVC. 

The receiving end decodes the video from HEVC to YUV. Qt only supports RGB32 so there is one more conversion from YUV to RGB32.

The small selfview does not use any compression. 

If your camera outputs YUV420 or RGB32, you don't need to modify anything else than the camera filter.

**Comment by Thanh-Binh** [comment]
well, thanks

**Comment by Thanh-Binh** [comment]
I tested SW by calling myself, it works.

But if I call another PC, I say PC1, application in PC1 crashed immerdiately. 
Here is the consol errors in PC1:
```QObject::connect: No such slot VideoSettings::initializeResolutions(QString)
QObject::connect:  (sender name:   'format_box')
QObject::connect:  (receiver name: 'VideoSettings')
QObject::connect: No such slot VideoSettings::initializeFramerates(QString)
QObject::connect:  (sender name:   'resolution')
QObject::connect:  (receiver name: 'VideoSettings')
QMetaObject::connectSlotsByName: No matching signal for on_about_clicked()
Settings, SIPSettings Resetting Advanced settings from UI
Settings, SIPSettings : Saving advanced Settings
This SIP connection uses an incoming connection: 1
Parsing SIP header with 9 lines
Found following SIP fields: ("Via", "Max-Forwards", "To", "From", "Call-ID", "CSeq", "Contact", "Content-Length")
Request detected: "INVITE"
Received a dialog creating INVITE. Creating dialog. CallID:  "QrQKKDLrWqPNPmV0" OurTag: "MEiWeJUpnSdlJIFd" Cseq: 1617094018
Incoming call, ConferenceView Displaying pop-up for somebody. Next layout location:  0 , 0
Getting local SDP suggestion
Generating new SDP message with our address as: "[::ffff:192.168.178.111]"
Sending, SIPContent : Composed SDP string: v=0
o=Binh 1617094021930 1617094021930 IN IP6 ::ffff:192.168.178.111
s=HEVC Video Call
c=IN IP6 ::ffff:192.168.178.111 
t=0 0
m=audio 0 RTP/AVP 0
a=sendrecv
m=video 0 RTP/AVP 97
a=rtpmap:97 h265/90000
a=sendrecv
a=candidate:1 1 UDP 2130706431 192.168.178.111 23099 typ host
a=candidate:1 2 UDP 2130706430 192.168.178.111 23100 typ host
a=candidate:1 3 UDP 2130706429 192.168.178.111 23101 typ host
a=candidate:1 4 UDP 2130706428 192.168.178.111 23102 typ host

Parsing SIP header with 9 lines
Found following SIP fields: ("Via", "Max-Forwards", "To", "From", "Call-ID", "CSeq", "Content-Length", "Content-Type")
Found media: "audio"
Correctly matched a flag attribute:  "sendrecv"
Found media: "video"
Correctly matched a flag attribute:  "sendrecv"
Parsed SDP is valid.
Successfully parsed SDP
Request detected: "ACK"
View, VideoFactory : Creating videowidget for sessionID: 1
[RTPLIB][WARNING][::set_payload] Unknown RTP format, setting clock rate to 10000
Settings for RGB32 to YUV threads: 1
"Kvazaar" iniating
Using preset ultrafast: --rd=0 --pu-depth-intra=2-3 --pu-depth-inter=1-2 --me=hexbs --gop=8 --ref=1 --bipred=1 --deblock=0:0 --signhide=0 --subme=0 --sao=off --rdoq=0 --rdoq-skip=0 --transform-skip=0 --mv-rdo=0 --full-intra-search=0 --smp=0 --amp=0 --cu-split-termination=zero --me-early-termination=sensitive --intra-rdo-et=0 --early-skip=1 --fast-residual-cost=28 --max-merge=5 
Initialization, Filter Getting custom Kvazaar options: 0
Compiled: INTEL, flags: MMX SSE SSE2
Detected: INTEL, flags: MMX SSE SSE2 SSE3 SSSE3 SSE41 SSE42 AVX
Available: sse2(2) sse41(4) 
In use: sse2(1) sse41(4) 
"Kvazaar" iniation succeeded.
Settings for RGB32 to YUV threads: 1
"Kvazaar" iniating
[hevc @ 0x55c6fd6921a0] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
[hevc @ 0x55c6fd6af3e0] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
Settings for YUV to RGB32 threads: 1
Processing, "Kvazaar" buffer full. Discarded input: 1 Total input: 4
[hevc @ 0x55c6fd6921a0] flush, decoder_0.
[hevc @ 0x7f30ec00ab70] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
[hevc @ 0x7f30ec00a530] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
Illegal instruction (core dumped)
```

The host PC has no crash with the following consol outputs:
```QObject::connect: No such slot VideoSettings::initializeResolutions(QString)
QObject::connect:  (sender name:   'format_box')
QObject::connect:  (receiver name: 'VideoSettings')
QObject::connect: No such slot VideoSettings::initializeFramerates(QString)
QObject::connect:  (sender name:   'resolution')
QObject::connect:  (receiver name: 'VideoSettings')
QMetaObject::connectSlotsByName: No matching signal for on_about_clicked()
Settings, SIPSettings Resetting Advanced settings from UI
Settings, SIPSettings : Saving advanced Settings
Local dialog created. CallID:  "G1HsG4gpx5VUZie2" Tag: "BciSHjZf5rzp2a8e" Cseq: 1617094608
SIP, Dialog client: Starting a call and sending an INVITE in session
Parsing SIP header with 8 lines
Found following SIP fields: ("Via", "Max-Forwards", "To", "From", "Call-ID", "CSeq", "Content-Length")
We don't yet have their remote Tag. Using the one in response.
Ringing, ConferenceView : Call is ringing. SessionID 1
Parsing SIP header with 10 lines
Found following SIP fields: ("Via", "Max-Forwards", "To", "From", "Call-ID", "CSeq", "Contact", "Content-Length", "Content-Type")
Found media: "audio"
Correctly matched a flag attribute:  "sendrecv"
Found media: "video"
Correctly matched a flag attribute:  "sendrecv"
Parsed SDP is valid.
Successfully parsed SDP
Sending, SIPContent : Composed SDP string: v=0
o=Binh 1617094611743 1617094611743 IN IP4 192.168.178.101
s=HEVCVideoCall
c=IN IP4 192.168.178.101 
t=0 0
m=audio 0 RTP/AVP 0
a=sendrecv
m=video 0 RTP/AVP 97
a=rtpmap:97 h265/90000
a=sendrecv
a=candidate:1 1 UDP 2130706431 192.168.178.101 23191 typ host
a=candidate:1 2 UDP 2130706430 192.168.178.101 23192 typ host
a=candidate:1 3 UDP 2130706429 192.168.178.101 23193 typ host
a=candidate:1 4 UDP 2130706428 192.168.178.101 23194 typ host

View, VideoFactory : Creating videowidget for sessionID: 1
[RTPLIB][WARNING][::set_payload] Unknown RTP format, setting clock rate to 10000
Settings for RGB32 to YUV threads: 1
"Kvazaar" iniating
Using preset ultrafast: --rd=0 --pu-depth-intra=2-3 --pu-depth-inter=1-2 --me=hexbs --gop=8 --ref=1 --bipred=1 --deblock=0:0 --signhide=0 --subme=0 --sao=off --rdoq=0 --rdoq-skip=0 --transform-skip=0 --mv-rdo=0 --full-intra-search=0 --smp=0 --amp=0 --cu-split-termination=zero --me-early-termination=sensitive --intra-rdo-et=0 --early-skip=1 --fast-residual-cost=28 --max-merge=5 
Initialization, Filter Getting custom Kvazaar options: 0
Compiled: INTEL, flags: MMX SSE SSE2
Detected: INTEL, flags: MMX SSE SSE2 SSE3 SSSE3 SSE41 SSE42 AVX AVX2
Available: avx2(54) sse2(2) sse41(4) 
In use: avx2(54) sse2(1) 
"Kvazaar" iniation succeeded.
Settings for RGB32 to YUV threads: 1
"Kvazaar" iniating
[hevc @ 0x55e342017140] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
[hevc @ 0x55e3420ba060] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
Settings for YUV to RGB32 threads: 1
Local dialog created. CallID:  "oGTvsDahmLAkohYD" Tag: "YYzvPLgNuaYW42q0" Cseq: 1617094612
SIP, Dialog client: Starting a call and sending an INVITE in session
```

Do you have any idea for solving this problem? Thanks

**Comment by polusto** [comment]
First make sure you are running the newest version, I see some errors that have been fixed some days ago.

I see illegal instruction caused the crash. That would indicate that either some library version such as kvazaar is not compiled for the correct platform or that some of the optimizations are not supported. 

Kvazzup doesn't yet detect which optimizations are supported. If you are using a machine that does not have AVX2, switch this line false :https://github.com/ultravideo/kvazzup/blob/207904cf143897d8a642ed6aa2ed9cd7864663ef/src/media/processing/yuvtorgb32.cpp#L25

Same is true SSE in the above line and also here : https://github.com/ultravideo/kvazzup/blob/207904cf143897d8a642ed6aa2ed9cd7864663ef/src/media/processing/rgb32toyuv.cpp#L12

**Comment by polusto** [comment]
I found an easy way to enable the optimizations automatically and updated the master:  81a1c0d134486809355dea2d6ebed356b7f4bb1b

If you still get the error after including the commit, make sure kvazzup has been compiled with correct optimizations.

**Comment by Thanh-Binh** [comment]
Yes, I have just downloaded the newest version, and turn off avx2.
But have the same crash
```Main:                                   Starting Kvazzup. (Location: /home/ubuntu/Desktop/VideoConf/bin)
QMetaObject::connectSlotsByName: No matching signal for on_about_clicked()
 ============================================================================= 
KvazzupController:                      Kvazzup initiation Started
 ============================================================================= 
ContactList:                            Reading contact list (# of contacts: 1)
VideoDrawHelper:                        Warning! Tried updating target rect before picture
Settings Helper:                        Warning! Found missing setting. Resetting video settings (Missing key: sip/ServerAddress)
Settings:                               Warning! Could not restore settings! Resetting settings from defaults.
Settings:                               Recording settings
VideoSettings:                          Resettings video settings fomr UI
VideoSettings:                          Initializing formats
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Saving video Settings
VideoSettings:                          Recording capability settings for device (Device Index: 0)
VideoSettings:                          Box status
                                        -- Format: -1
                                        -- Resolution: -1

Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Recorded following video settings.
                                        -- Resolution: 0x0
                                        -- Resolution Index: 0
                                        -- Format: No_camera

Settings Helper:                        Writing list from GUI to settings
                                        -- File: kvazzup.ini
                                        -- List name: parameters
                                        -- Table items: 0

AudioSettings:                          Saving audio Settings
Settings:                               Opening settings
Settings Helper:                        ERROR! Missing setting found (Key: video/OWF)
AudioSettings:                          Saving audio Settings
VideoSettings:                          Initializing formats
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Max Threads (Threads: 4)
Settings Helper:                        ERROR! Missing setting found (Key: video/OWF)
VideoSettings:                          Resettings video settings fomr UI
VideoSettings:                          Initializing formats
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Saving video Settings
VideoSettings:                          Recording capability settings for device (Device Index: 0)
VideoSettings:                          Box status
                                        -- Format: -1
                                        -- Resolution: -1

Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Recorded following video settings.
                                        -- Resolution: 0x0
                                        -- Resolution Index: 0
                                        -- Format: No_camera

Settings Helper:                        Writing list from GUI to settings
                                        -- File: kvazzup.ini
                                        -- List name: parameters
                                        -- Table items: 0

Settings Helper:                        Reading list from settings
                                        -- File: blocklist.local
                                        -- List name: blocklist
                                        -- Items: 0

CallWindow:                             Initiating, CallWindow: Creating statistics window
ChartPainter:                           Initiating chart (Title: Bandwidth (kbit/s))
ChartPainter:                           Initiating chart (Title: Bit rates (kbit/s))
ChartPainter:                           Initiating chart (Title: Bit rates (kbit/s))
ChartPainter:                           Initiating chart (Title: Latencies (ms))
ChartPainter:                           Initiating chart (Title: Latencies (ms))
ChartPainter:                           Initiating chart (Title: Frame rates (fps))
SIPManager:                             Listening to SIP TCP connections (Port: 5060)
SIPDialogManager:                       Iniating SIP Transactions
SIPRegistrations:                       Initiatin Registrations
NetworkCandidates:                      Allocating (but not reserving) ports for media
                                        -- Min ports: 38500
                                        -- Max port: 39500

MediaManager:                           Initiating
FilterGraph:                            Iniating camera and selfview
CameraFilter:                           Initiating Qt camera (ID: 0)
CameraFilter:                           Using the following camera settings.
                                        -- Format: No_camera
                                        -- Resolution: 640x480
                                        -- Framerate: 30 to 30

ChartPainter:                           Initiating chart (Title: Frame rates (fps))
 ============================================================================= 
CameraFilter:                           Starting camera.
 ============================================================================= 
FilterGraph:                            Connecting filters (Connection: Camera -> Display)
FilterGraph:                            Connecting filters (Connection: Screen Sharing -> Display)
 ============================================================================= 
KvazzupController:                      Kvazzup initiation finished
 ============================================================================= 
VideoDrawHelper:                        Warning! Tried updating target rect before picture
GUIMessage:                             Closing message (Waiting messages: 0)
Settings:                               Closing Settings. Gettings settings from file.
Settings Helper:                        Warning! Found missing setting. Resetting video settings (Missing key: sip/ServerAddress)
Settings:                               Recording settings
VideoSettings:                          Saving video settings
VideoSettings:                          Saving video Settings
VideoSettings:                          Recording capability settings for device (Device Index: 0)
VideoSettings:                          Box status
                                        -- Format: -1
                                        -- Resolution: -1

Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Recorded following video settings.
                                        -- Resolution: 0x0
                                        -- Resolution Index: 0
                                        -- Format: No_camera

Settings Helper:                        Writing list from GUI to settings
                                        -- File: kvazzup.ini
                                        -- List name: parameters
                                        -- Table items: 0

CameraFilter:                           Updating camera settings since they have changed
CameraFrameGrabber:                     Warning! Not a valid frame
CameraFilter:                           Initiating Qt camera (ID: 0)
CameraFilter:                           Using the following camera settings.
                                        -- Format: No_camera
                                        -- Resolution: 640x480
                                        -- Framerate: 30 to 30

ChartPainter:                           Initiating chart (Title: Frame rates (fps))
 ============================================================================= 
CameraFilter:                           Starting camera.
 ============================================================================= 
FilterGraph:                            Starting camera
Common:                                 Warning! Found faulty setting (Key: sip/ServerAddress)
SIPManager:                             Warning! SIP Registrar was empty or we have already registered. No registering.
VideoSettings:                          Cancelled modifying video settings. Getting settings from system.
VideoSettings:                          Initializing formats
Camera Info:                            Getting text of 0 video formats (DeviceID: 0)
VideoSettings:                          Max Threads (Threads: 4)
VideoSettings:                          Restoring previous video settings from file. (Filename: /home/ubuntu/Desktop/VideoConf/bin/kvazzup.ini)
Settings Helper:                        Reading list from settings
                                        -- File: kvazzup.ini
                                        -- List name: parameters
                                        -- Items: 0

Settings:                               Opening settings
Settings:                               Closing Settings. Gettings settings from file.
Settings Helper:                        Warning! Found missing setting. Resetting video settings (Missing key: sip/ServerAddress)
Settings:                               Closing Settings. Gettings settings from file.
Settings Helper:                        Warning! Found missing setting. Resetting video settings (Missing key: sip/ServerAddress)
KvazzupController:                      Starting call with contact (Contact: BinhHP)
SIPTransport:                           Initiating TCP connection for sip connection (TransportID: 1)
TCPConnection:                          Establishing connection
                                        -- Destination: 192.168.178.101
                                        -- port: 5060

 ============================================================================= 
TCPConnection:                          Starting TCP loop
 ============================================================================= 
TCPConnection:                          Starting to connect TCP
TCPConnection:                          Checking status (Address: 192.168.178.101:5060)
TCPConnection:                          Attempting to connect (Attempt: 1)
TCPConnection:                          Connected succesfully (Connection: 192.168.178.111:37948 <-> 192.168.178.101:5060)
SIPDialogManager:                       Intializing a new dialog by sending an INVITE
SIPDialogState:                         Creating a new dialog.
Local dialog created. CallID:  "FUjXaXS1huWMxnHC" Tag: "juw5B7nF7nJXr9JP" Cseq: 1617104989
SIPDialogState:                         Setting peer-to-peer address.
SIP, Dialog client: Starting a call and sending an INVITE in session
SIPDialogClient:                        Client starts sending a request. (Type: 1)
SIPDialog:                              Iniate sending of a dialog request
SIPDialogState:                         Increasing CSeq (CSeq: 1617104990)
 ============================================================================= 
SIPTransport:                           Composing and sending SIP Request: (Type: INVITE)
 ============================================================================= 
SIPDialog:                              Finished sending of a dialog request
TCPConnection:                          Writing buffer to TCP socket (Buffer size: 1)
ConferenceView:                         Adding widget to display that we are calling someone. (SessionID: 1)
ConferenceView:                         Checks session: SessionID does not exist
ConferenceView:                         Checks session: SessionID does not exist
ConferenceView:                         Initializing session (SessionID: 1)
TCPConnection:                          Written to socket (Bytes: 382)
TCPConnection:                          Can read one line (Bytes available: 318)
SIPTransport:                           Parsing package to header and body
                                        -- Messages parsed so far: 0
                                        -- Header end index: 318
                                        -- content-length index: 297

SIPTransport:                           Parsed Content-length (Content-length: 0)
Parsing SIP header with 8 lines
Found following SIP fields: ("Via", "Max-Forwards", "To", "From", "Call-ID", "CSeq", "Content-Length")
 ============================================================================= 
SIPTransport:                           Parsing incoming response (Type: 180)
 ============================================================================= 
SIPRouting:                             Found our via. This is meant for us!
SIPDialogManager:                       Starting to process identifying SIP response dialog.
We don't yet have their remote Tag. Using the one in response.
SIPDialogManager:                       Found dialog matching the response
SIPDialogClient:                        Client starts processing response
SIPDialogClient:                        Got a provisional response. Restarting timer.
KvazzupController:                      Our call is ringing
Ringing, ConferenceView : Call is ringing. SessionID 1
SIPDialogManager:                       Response processing finished (SessionID: 1)
TCPConnection:                          Can read one line (Bytes available: 887)
SIPTransport:                           Parsing package to header and body
                                        -- Messages parsed so far: 0
                                        -- Header end index: 411
                                        -- content-length index: 357

SIPTransport:                           Parsed Content-length (Content-length: 476)
Parsing SIP header with 10 lines
Found following SIP fields: ("Via", "Max-Forwards", "To", "From", "Call-ID", "CSeq", "Contact", "Content-Length", "Content-Type")
Found media: "audio"
Correctly matched a flag attribute:  "sendrecv"
Found media: "video"
Correctly matched a flag attribute:  "sendrecv"
Parsed SDP is valid.
Successfully parsed SDP
 ============================================================================= 
SIPTransport:                           Parsing incoming response (Type: 200)
 ============================================================================= 
SIPRouting:                             Found our via. This is meant for us!
SIPDialogManager:                       Starting to process identifying SIP response dialog.
SIPDialogManager:                       Found dialog matching the response
SIPManager:                             Got first SDP offer.
Negotiation:                            Found following codecs in SDP (Codecs: pcm, h265)
SDPNegotiator:                          Found suitable RTP number.
SDPNegotiator:                          Found suitable codec.
ICE:                                    Start Generating ICE candidates
                                        -- Local: 4
                                        -- Global: 0
                                        -- STUN: 0
                                        -- STUN relays: 0
                                        -- TURN: 0

 ============================================================================= 
ICE:                                    Starting ICE nomination
 ============================================================================= 
ICE:                                    Created 4 candidate pairs
SIPDialogClient:                        Client starts processing response
 ============================================================================= 
KvazzupController:                      They accepted our call!
 ============================================================================= 
SIPDialogClient:                        Client starts sending a request. (Type: 2)
SIPDialog:                              Iniate sending of a dialog request
SIPManager:                             Adding SDP content to request
 ============================================================================= 
SIPTransport:                           Composing and sending SIP Request: (Type: ACK)
 ============================================================================= 
SIPContent:                             Parameter SDP is valid. Starting to compose to string.
Sending, SIPContent : Composed SDP string: v=0
o=BinhT420 1617104991326 1617104991326 IN IP4 192.168.178.111
s=HEVCVideoCall
c=IN IP4 192.168.178.111 
t=0 0
m=audio 0 RTP/AVP 0
a=sendrecv
m=video 0 RTP/AVP 97
a=rtpmap:97 h265/90000
a=sendrecv
a=candidate:1 1 UDP 2130706431 192.168.178.111 38500 typ host
a=candidate:1 2 UDP 2130706430 192.168.178.111 38501 typ host
a=candidate:1 3 UDP 2130706429 192.168.178.111 38502 typ host
a=candidate:1 4 UDP 2130706428 192.168.178.111 38503 typ host

SIPDialog:                              Finished sending of a dialog request
TCPConnection:                          Writing buffer to TCP socket (Buffer size: 1)
IcePairTester:                          Controller starts connectivity tests. (Pair: 192.168.178.111:38500 <-> 192.168.178.101:38500)
KvazzupController:                      Call negotiated
KvazzupController:                      Waiting for the ICE to finish before starting the call.
SIPDialogManager:                       Response processing finished (SessionID: 1)
IcePairTester:                          Controller starts connectivity tests. (Pair: 192.168.178.111:38502 <-> 192.168.178.101:38502)
TCPConnection:                          Written to socket (Bytes: 842)
IcePairTester:                          Controller starts connectivity tests. (Pair: 192.168.178.111:38503 <-> 192.168.178.101:38503)
IcePairTester:                          Controller starts connectivity tests. (Pair: 192.168.178.111:38501 <-> 192.168.178.101:38501)
IcePairTester:                          Controller found a connection (Pair: 192.168.178.111:38500 <-> 192.168.178.101:38500)
IceSessionTester:                       Controller component finished (Finished components: 1/4)
IcePairTester:                          Controller found a connection (Pair: 192.168.178.111:38503 <-> 192.168.178.101:38503)
IcePairTester:                          Controller found a connection (Pair: 192.168.178.111:38501 <-> 192.168.178.101:38501)
IcePairTester:                          Controller found a connection (Pair: 192.168.178.111:38502 <-> 192.168.178.101:38502)
IceSessionTester:                       Controller component finished (Finished components: 2/4)
IceSessionTester:                       Controller component finished (Finished components: 3/4)
IceSessionTester:                       Controller component finished (Finished components: 4/4)
 ============================================================================= 
ICE:                                    ICE finished.
                                        -- Component 1: 192.168.178.111:38500 <-> 192.168.178.101:38500
                                        -- Component 2: 192.168.178.111:38501 <-> 192.168.178.101:38501
                                        -- Component 3: 192.168.178.111:38502 <-> 192.168.178.101:38502
                                        -- Component 4: 192.168.178.111:38503 <-> 192.168.178.101:38503

 ============================================================================= 
Negotiation:                            ICE nomination has succeeded (SessionID: 1)
KvazzupController:                      ICE has been successfully completed (SessionID: 1)
KvazzupController:                      Call has been agreed upon with peer. (SessionID: 1)
ConferenceView:                         Adding Videostream. (SessionID: 1)
View, VideoFactory : Creating videowidget for sessionID: 1
VideoviewFactory:                       Created video widget.
                                        -- SessionID: 1
                                        -- View Number: 1

ConferenceView:                         Clearing all previous views.
VideoDrawHelper:                        Warning! Tried updating target rect before picture
Delivery:                               Adding new peer
Delivery:                               RTP streamer (SessionID: 1)
MediaManager:                           Start creating media.
                                        -- Outgoing media: 2
                                        -- Incoming media: 2

MediaManager:                           Using media specific address for outgoing media.
                                        -- Type: audio
                                        -- Path: 192.168.178.111:38502 -> 192.168.178.101:38502

Delivery:                               Creating mediastream
Delivery:                               Creating sender filter
AudioCaptureFilter:                     Initializing audio capture filter.
[RTPLIB][WARNING][::set_payload] Unknown RTP format, setting clock rate to 10000
AudioCaptureFilter:                     A microphone chosen. (Device name: default)
AudioCaptureFilter:                     Creating audio input. (Notify interval (ms): 1000)
AudioCaptureFilter:                     Audio initializing completed.
AudioCaptureFilter:                     Resuming audio input.
AECProcessor:                           Initiating echo frame processing (Filter length: 1600)
FilterGraph:                            Connecting filters (Connection: Audio_Capture -> AEC input)
FilterGraph:                            Starting microphone
AudioCaptureFilter:                     Resuming audio input.
MediaManager:                           Using media specific address for outgoing media.
                                        -- Type: video
                                        -- Path: 192.168.178.111:38500 -> 192.168.178.101:38500

Delivery:                               Creating mediastream
Delivery:                               Creating sender filter
UvgRTPSender:                           Updated buffersize (Size: 64)
FilterGraph:                            Adding send video (SessionID: 1)
FilterGraph:                            Iniating video send
FilterGraph:                            Filter output and input do not match. Finding conversion (Connection: Camera->Kvazaar)
FilterGraph:                            Adding RGB32 to YUV conversion
FilterGraph:                            Connecting filters (Connection: Camera -> RGB32toYUV)
FilterGraph:                            Connecting filters (Connection: RGB32toYUV -> Kvazaar)
"Kvazaar" iniating
Using preset ultrafast: --rd=0 --pu-depth-intra=2-3 --pu-depth-inter=1-2 --me=hexbs --gop=8 --ref=1 --bipred=1 --deblock=0:0 --signhide=0 --subme=0 --sao=off --rdoq=0 --rdoq-skip=0 --transform-skip=0 --mv-rdo=0 --full-intra-search=0 --smp=0 --amp=0 --cu-split-termination=zero --me-early-termination=sensitive --intra-rdo-et=0 --early-skip=1 --fast-residual-cost=28 --max-merge=5 
Initialization, Filter Getting custom Kvazaar options: 0
Compiled: INTEL, flags: MMX SSE SSE2
Detected: INTEL, flags: MMX SSE SSE2 SSE3 SSSE3 SSE41 SSE42 AVX
Available: sse2(2) sse41(4) 
In use: sse2(1) sse41(4) 
"Kvazaar" iniation succeeded.
FilterGraph:                            Filter output and input do not match. Finding conversion (Connection: Screen Sharing->Kvazaar)
FilterGraph:                            Adding RGB32 to YUV conversion
FilterGraph:                            Connecting filters (Connection: Screen Sharing -> RGB32toYUV)
FilterGraph:                            Connecting filters (Connection: RGB32toYUV -> Kvazaar)
"Kvazaar" iniating
MediaManager:                           Using media specific address for incoming.
                                        -- Type: audio
                                        -- Address: 192.168.178.111
                                        -- Port: 38502

Delivery:                               Creating receiver filter
FilterGraph:                            Connecting filters (Connection: RTP Receiver PCM -> Audio Mixer)
MediaManager:                           Using media specific address for incoming.
                                        -- Type: video
                                        -- Address: 192.168.178.111
                                        -- Port: 38500

Delivery:                               Creating receiver filter
FilterGraph:                            Connecting filters (Connection: RTP Receiver HEVC -> OpenHEVC)
Filter:                                 Starting to initiate OpenHEVC
[hevc @ 0x55ef09981e30] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
[hevc @ 0x55ef090434f0] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
Filter:                                 OpenHEVC initiation successful. (Version: OpenHEVC v2.0)
FilterGraph:                            Filter output and input do not match. Finding conversion (Connection: OpenHEVC->Display)
FilterGraph:                            Adding YUV to RGB32 conversion
FilterGraph:                            Connecting filters (Connection: OpenHEVC -> YUVtoRGB32)
FilterGraph:                            Connecting filters (Connection: YUVtoRGB32 -> Display)
KvazzupController:                      Waiting for the ICE to finish before starting the call.
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioCaptureFilter:                     Audio Input State changed (States:: Current: 0, Wanted: 0)
Processing, "Kvazaar" buffer full. Discarded input: 1 Total input: 4
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
AudioOutputDevice:                      Warning! Read too little audio data. (Read vs available: 0 vs 1280)
Filter:                                 Parameter set found (Frame received before: 0)
Filter:                                 PEER ERROR:Detected no slices in incoming stream.
Filter:                                 Uniniating.
[hevc @ 0x55ef09981e30] flush, decoder_0.
Filter:                                 Starting to initiate OpenHEVC
[hevc @ 0x7fde20076530] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
[hevc @ 0x7fde20035f00] nb threads_frame = 1, nb threads_slice 1, thread_type = null 
Filter:                                 OpenHEVC initiation successful. (Version: OpenHEVC v2.0)
Illegal instruction (core dumped)

```

**Comment by Thanh-Binh** [comment]
Moreover, in the newest version I can not save and close VideoSettings.
By Close VideoSetting, the VideoSetting window is NOT closed/disappeared

**Comment by polusto** [comment]
I don't know what is causing that Illegal instruction. Can you run this in debugger and paste result?

I am unable to replicate the settings issue. For some reason it seems to complain about not having the serverAddress set. Can you set the server address (anything will do) and does this fix the issue?

**Comment by Thanh-Binh** [comment]
I use a server adress like
sip:BinhH@192.168.178.111
What is your address in your test? Maybe I can test directly by calling you

**Comment by polusto** [comment]
Ok, in that case I'm not sure what is causing the video settings not closing. Is it still giving that error? I fixed the OWF error in master.

In order to call another person a proxy is needed and I'm currently working on getting these kinds of connections working with authentication support. In other words, current master cannot bypass firewalls or NATs with SIP.

Another problem is that these issues may be related to Linux and currently I don't have a Linux computer to test with.

If you are using Qt creator, you can run (if it is set up correctly) debugger using 
![image](https://user-images.githubusercontent.com/6498638/113090730-15937d80-91f3-11eb-8cab-a5afef94c94d.png) symbol in debug mode.

**Comment by Thanh-Binh** [comment]
Could you please send me your complete runable binary compiled for windows , so that I can call from windows to Linux?
Problem is that only one in a Linux is crashed!

**Comment by polusto** [comment]
I don't understand why you need windows binary if Linux version crashed?

Also, why can't you run the crashing Linux version in debugger? That would tell immediately what causes the crash.

**Comment by Thanh-Binh** [comment]
Since it crashed only in a non-development Linux. 
Development-Linux runs without any problem, so that I think it may be a Linux problem, so that the test with standalone app in windows will help us better to understand !

that is why I can not identify the crash even in debugger mode!

**Comment by polusto** [comment]
I created a new issue #6  for this illegal instruction crash. We should continue the discussion there. You may also open new issues if something else is still causing problems.
